README.md
# canon-d

A Clojure library designed to ... well, that part is up to you.

## Usage

FIXME

## License

Copyright © 2022 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2022-04-28
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2022-04-28
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/canon-d/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/canon-d/compare/0.1.0...0.1.1


test/canon_d/core_test.clj
(ns canon-d.core-test
  (:require [clojure.test :refer :all]
            [canon-d.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/canon_d/core.clj
(ns canon-d.core
  (:require [overtone.live :as ol]))

(ol/demo (ol/saw 440))
(ol/stop)

(require '[overtone.inst.sampled-piano :as p])

(p/sampled-piano (ol/note :c4))
(p/sampled-piano (ol/note :d4))
(p/sampled-piano (ol/note :e4))
(def piano-sound-f4
  (p/sampled-piano (ol/note :f4)))
(ol/ctl piano-sound-f4 :amp 0.1 :note (ol/note :c4))

(def dirty-kick
  (ol/sample "~/Downloads/drum/OH/kick_OH_P_2.wav"))
(def semi-open-hat
  (ol/sample
   "/home/gzmask/Downloads/drum/OH/hihatSemiOpen5_OH_P_1.wav"))
(ol/defsample semi-open-hat2 "/home/gzmask/Downloads/drum/OH/hihatSemiOpen5_OH_P_2.wav")


(def ring-hat (ol/freesound 12912))
(def snare (ol/freesound 26903))
(def click (ol/freesound 406))
(def wop (ol/freesound 85291))
(def subby (ol/freesound 25649))
(dirty-kick :amp 1.0)
(semi-open-hat)
(semi-open-hat2)
(ring-hat)
(snare)
(click)
(click :decay 0.05 :amp 0.3)
(wop :amp 1.0)
(subby)

(def m (ol/metronome 128))

(defn player
  [beat]
  (let [next-beat (inc beat)]
    (when (zero? (mod beat 16))
      (ol/at (m beat) (p/sampled-piano (ol/note :c4) 0.2))
      (ol/at (m (+ beat 0.1)) (p/sampled-piano (ol/note :e3) 0.1))
      (ol/at (m (+ beat 0.2)) (p/sampled-piano (ol/note :g4) 0.1))
      (ol/at (m (+ beat 2)) (p/sampled-piano (ol/note :g3) 0.2))
      (ol/at (m (+ beat 2.1)) (p/sampled-piano (ol/note :b3) 0.1))
      (ol/at (m (+ beat 2.2)) (p/sampled-piano (ol/note :d3) 0.1))
      (ol/at (m (+ beat 4)) (p/sampled-piano (ol/note :a4) 0.2))
      (ol/at (m (+ beat 4.1)) (p/sampled-piano (ol/note :c4) 0.1))
      (ol/at (m (+ beat 4.2)) (p/sampled-piano (ol/note :e4) 0.1)))
    (when (zero? (mod beat 4))
      (ol/at (m beat) (dirty-kick :amp 0.8))
      (ol/at (m (+ beat 1/4)) (dirty-kick :amp 0.8))
      (ol/at (m (+ beat 1)) (wop :amp 0.8))
      (ol/at (m (+ beat 2)) (subby :amp 1.0))
      (ol/at (m (+ beat 3)) (dirty-kick :amp 0.8))
      (ol/at (m (+ beat 3.5)) (dirty-kick :amp 0.8)))
    (ol/at (m beat) (click :amp 0.2))
    (ol/at (m (+ beat 1/2)) (dirty-kick :amp 0.3))
    (ol/at (m (+ beat 2/3)) (wop :amp 0.1))
    (ol/at (m (+ beat 3/4)) (subby :amp 0.1))
    (ol/apply-by (m next-beat) #'player [next-beat])))

(player (m))
(ol/stop)

(ol/definst guitar-solo [] (ol/sound-in [0 1]))
(def guitar-solo-track (guitar-solo :amp 0.5))
(ol/ctl guitar-solo-track :amp 0.8)
(ol/stop)

(ol/recording-start "guitar.wav")
(ol/recording-stop)
;(def guitar-solo (ol/sample "/home/gzmask/project/canon-d/guitar.wav"))
(ol/defsample guitar-solo "/home/gzmask/project/canon-d/guitar.wav" :start (* 4 (ol/server-sample-rate)))
(guitar-solo :amp 10.0)
(ol/free-all-loaded-samples)
(ol/free-sample guitar-solo)


src/canon_d/raining_car.clj
(ns canon-d.raining-car
  (:require [overtone.core :as o]))

(def internal-server
  (o/boot-internal-server))

(o/demo (o/saw 440))
(o/stop)

(meta #'o/demo)

(require '[overtone.inst.sampled-piano :as p])

(p/sampled-piano (o/note :A3))

(def dirty-kick (o/freesound 344757 :size (int (* 1.1 (o/server-sample-rate)))))
(def ring-hat (o/freesound 387347))
(def snare (o/freesound 26903))
(def click (o/freesound 406))
(def wop (o/freesound 85291))
(def subby (o/freesound 25649))
(dirty-kick)
(ring-hat)
(snare)
(click)
(wop)
(subby)

(o/buffer-info dirty-kick)
(o/buffer-info ring-hat)
(o/buffer-info subby)

(meta #'o/sample-rate)
(o/sample-rate:ir)

(require '[overtone.inst.synth :as sth])

(sth/overpad (o/note :A3))

(def a-buffer (o/buffer (* 5 (o/server-sample-rate))))
(o/free-sample a-buffer)
(o/definst go-recorder [freq 440 attack 0.01 sustain 0.4 release 0.1 vol 0.4]
  (let [aud-in (o/sound-in)
        aud-in (o/record-buf:ar aud-in a-buffer)
        aud-out (* 10 vol aud-in)]
    (o/pan2:ar aud-out)))
(o/inst-fx! go-recorder o/fx-reverb)
(o/clear-fx go-recorder)
(go-recorder)
(o/stop)

(o/definst go-buffer [freq 440 attack 0.01 sustain 0.4 release 0.1 vol 1]
  (o/pan2 (* 10 vol (o/play-buf:ar 1 a-buffer))))
(o/inst-fx! go-buffer o/fx-reverb)
(o/clear-fx go-buffer)
(go-buffer)
(o/buffer-info a-buffer)
(o/buffer-save a-buffer "123.wav")
(meta #'o/play-buf)
(meta #'o/record-buf)


(o/defsynth pitch-controlled-saws
  [out-bus 0]
  (let [p   (o/pitch (o/sound-in))
        p   (/ p 4)
        p   (o/lag p 1)
        s   (o/saw (first p))
        s   (o/record-buf:ar (* 0.4 s) a-buffer)]
    (o/out out-bus (o/pan2 s))))
(pitch-controlled-saws)
(o/stop)

(def m (o/metronome 128))

(defn get-ms-per-beat [m]
  (int (- (m 2) (m 1))))

(def ms-per-beat (get-ms-per-beat m))

(require '[overtone.at-at :as at])

(def my-pool (at/mk-pool))

(at/at (+ 1000 (o/now)) #(dirty-kick :amp 0.2) my-pool)

(defn base-drums [b]
  (at/at (m b) #(dirty-kick :amp 0.2) my-pool)
  (at/at (m (+ 0.5 b)) #(subby :amp 0.2) my-pool)
  (at/at (m (inc b)) #(wop :amp 0.2) my-pool)
  (at/at (m (+ 3 b)) #(click :amp 0.2) my-pool))
(base-drums (m))

(defn progressing [b]
  (doseq [t (range 0 2 0.25)]
    (at/at (m (+ t b)) #(dirty-kick :amp (* 0.5 (rand 1))) my-pool)))
(progressing (m))

(defn base-line [b]
  (doall
   (map
    (fn [note t]
      (at/at (m (+ t b)) #(sth/overpad note :amp (rand-nth [0 0.1 0.2])) my-pool))
    (rand-nth [(shuffle (o/scale :a3 :minor))
               (conj (shuffle (o/scale :a3 :minor))
                     (rand-nth [64 69]))])
    (range 0 9 (rand-nth [1 0.5 0.25])))))
(base-line (m))

(defn player [b]
  (let [next-beat (inc b)]
    (when (zero? (mod b 4))
      (base-drums b))
    (when (zero? (mod b 8))
      (progressing b))
    (when (zero? (mod b 16))
      (base-line b))
    (at/at (m next-beat) #(#'player next-beat) my-pool)))

(player (m))
(o/stop)
(at/scheduled-jobs my-pool)
(at/stop-and-reset-pool! my-pool :strategy :kill)



src/canon_d/promise.clj
(ns canon-d.promise
  (:require [overtone.core :as o]))

(def internal-server
  (o/boot-internal-server))

(o/demo (o/pan2 (o/sin-osc 440) (o/line -1 1 2)))

(o/defsynth dubstep [bpm 120 wobble 1 freq 440 snare-vol 1 kick-vol 1 v 1 out-bus 0]
 (let [trig (o/impulse:kr (/ bpm 120))
       swr (o/demand trig 0 (o/dseq [wobble] o/INF))
       sweep (o/lin-exp (o/lf-tri swr) -1 1 40 3000)
       wob (apply + (o/saw (* freq [0.99 1.01])))
       wob (o/lpf wob sweep)
       wob (* 0.8 (o/normalizer wob))
       wob (+ wob (o/bpf wob 1500 2))
       wob (+ wob (* 0.2 (o/g-verb wob 9 0.7 0.7)))

       kickenv (o/decay (o/t2a (o/demand (o/impulse:kr (/ bpm 30)) 0 (o/dseq [1 0 0 0 0 0 1 0 1 0 0 1 0 0 0 0] o/INF))) 0.7)
       kick (* (* kickenv 7) (o/sin-osc (+ 40 (* kickenv kickenv kickenv 200))))
       kick (o/clip2 kick 1)

       snare (* 3 (o/pink-noise) (apply + (* (o/decay (o/impulse (/ bpm 240) 0.5) [0.4 2]) [1 0.05])))
       snare (+ snare (o/bpf (* 4 snare) 2000))
       snare (o/clip2 snare 1)]

   (o/out out-bus (* v (o/clip2 (+ wob (* kick-vol kick) (* snare-vol snare)) 1)))))

(comment
  (def d (dubstep))
  (o/ctl d :wobble 2)
  (o/ctl d :freq (o/midi->hz (o/note :d4)))
  (o/ctl d :bpm 120)
  (o/stop)
  )

(o/odoc o/in)
(o/odoc o/sound-in)

(require '[overtone.inst.sampled-piano :as p])

(p/sampled-piano (o/note :d3))
(p/sampled-piano (o/note :e3))
(p/sampled-piano (o/note :f3))
(p/sampled-piano (o/note :g3))
(p/sampled-piano (o/note :a3))
(p/sampled-piano (o/note :a#3))
(p/sampled-piano (o/note :c4))
(p/sampled-piano (o/note :d4))


(map o/find-note-name (o/scale :d3 :minor))

(def m (o/metronome 120))
(- (m 2) (m 1))
(doseq [note (map o/find-note-name (concat (o/scale :d3 :minor)
                                           (rest (o/scale :d4 :minor))))]
  (p/sampled-piano (o/note note))
  (Thread/sleep 500))

(require '[overtone.at-at :as at])

(def beat-pool (at/mk-pool))

(at/at (m) #(p/sampled-piano (o/note :d3)) beat-pool)

(defmacro at [metronome play-fn]
  `(at/at ~metronome #(~play-fn) beat-pool))

(at (m (+ 2 (m))) (p/sampled-piano (o/note :d3)))
(defn intro1 [b]
  (at (m b) (p/sampled-piano (o/note :a3) :level 0.5))
  (at (m (+ 1 b)) (p/sampled-piano (o/note :d3) :level 0.5))
  (at (m (+ 2 b)) (p/sampled-piano (o/note :f4) :level 0.5))
  (at (m (+ 3 b)) (p/sampled-piano (o/note :a3) :level 0.5)))
(intro1 (m))
(defn intro2 [b]
  (at (m (+ 0 b)) (p/sampled-piano (o/note :a3) :level 0.5))
  (at (m (+ 1 b)) (p/sampled-piano (o/note :d3) :level 0.5))
  (at (m (+ 2 b)) (p/sampled-piano (o/note :f4) :level 0.5))
  (at (m (+ 3 b)) (p/sampled-piano (o/note :a#3) :level 0.5)))
(intro2 (m))
(defn intro3 [b]
  (at (m (+ 0 b)) (p/sampled-piano (o/note :g3) :level 0.5))
  (at (m (+ 1 b)) (p/sampled-piano (o/note :a#3) :level 0.5))
  (at (m (+ 2 b)) (p/sampled-piano (o/note :f4) :level 0.5))
  (at (m (+ 3 b)) (p/sampled-piano (o/note :a#3) :level 0.5)))
(intro3 (m))
(defn intro [b]
  (intro1 b)
  (intro1 (+ 8 b))
  (intro2 (+ 16 b))
  (intro3 (+ 24 b))
  (at (m (+ 32 b)) (#'intro (+ 32 b))))
(intro (m))

(require '[overtone.inst.synth :as sth])
(sth/overpad (o/note :a3))
(sth/overpad (o/note :g3))
(defn base-line-one [b]
  (doall
   (map (fn [n b]
          (at (m b) (sth/overpad (o/note n)))
          (at (m (+ 1 b)) (sth/overpad (o/note n) :amp 1.1))
          (at (m (+ 2 b)) (sth/overpad (o/note n) :amp 1.1))
          (at (m (+ 3 b)) (sth/overpad (o/note n) :amp 1.1)))
        [:d3 :e3 :g3]
        [b (+ b 4) (+ b 8)])))
(defn base-line [b]
  (base-line-one b)
  (at (m (+ b 12)) (#'base-line (+ b 12))))
(base-line (m))

(require '[overtone.inst.drum :as d])

(d/hat3)
(d/clap)
(d/noise-snare)
(d/open-hat)
(d/quick-kick)
(d/dance-kick)
(d/snare)
(d/tom)

(defn drums-pat [b]
  (at (m (+ 0 b)) (d/hat3 :amp 1.0))
  (at (m (+ 1 b)) (d/snare :amp 1.0))
  (at (m (+ 2 b)) (d/snare :amp 1.0))
  (at (m (+ 3 b)) (d/quick-kick :amp 1.0))
  (at (m (+ 4 b)) (d/quick-kick :amp 0.8))
  (at (m (+ 5 b)) (d/clap :amp 0.8))
  (at (m (+ 5 b)) (d/snare :amp 0.8))
  (at (m (+ 7 b)) (d/quick-kick :amp 0.8)))

(defn drums [b]
  (drums-pat b)
  (at (m (+ b 8)) (#'drums (+ b 8))))
(drums (m))

(defn rec-guitar [reusable-long-buffer vol]
  (let [p (o/definst guitar-rec [vol 0.5]
            (let [guitar-in (o/sound-in [0 1])
                  guitar-out (o/record-buf:ar guitar-in reusable-long-buffer)]
              (* 10 vol guitar-out)))
        start-time (o/now)
        synth-id (p)]
    (fn stop-recording []
      (let [_ (o/kill synth-id)
            stop-time (o/now)
            buffer-length-ms (- stop-time start-time)
            buffer-length-seconds (/ buffer-length-ms 1000)
            out-buffer (o/buffer (* buffer-length-seconds
                                    (o/server-sample-rate)) 2)
            p (o/definst guitar-rec [vol 0.5]
                (let [guitar-in (o/play-buf:ar 2 reusable-long-buffer)
                      guitar-out (o/record-buf:ar guitar-in out-buffer)]
                  (* 10 vol guitar-out)))
            pid (p)]
        (at (+ (o/now) buffer-length-ms) (o/kill pid))
        out-buffer))))

(defn play-guitar [out-buffer]
  (o/definst go-buffer [vol 1]
    (* 10 vol (o/play-buf:ar 2 out-buffer)))
  (go-buffer))

(def buffer-pool (o/buffer (* 30 (o/server-sample-rate)) 2))
(def g15-73-15-37-stop-recod (rec-guitar buffer-pool 1.0))
(def g15-73-15-37-buffer (g15-73-15-37-stop-recod))
(play-guitar g15-73-15-37-buffer)
(def g77765-44543-77717-332-stop-rec (rec-guitar buffer-pool 1.0))
(def g77765-44543-77717-332-buffer (g77765-44543-77717-332-stop-rec))
(play-guitar g77765-44543-77717-332-buffer)
(def g3-21-123-4345-stop-rec (rec-guitar buffer-pool 1.0))
(def g3-21-123-4345-buffer (g3-21-123-4345-stop-rec))
(play-guitar g3-21-123-4345-buffer)

(o/stop)
(o/free-all-loaded-samples)
(at/stop-and-reset-pool! beat-pool :strategy :kill)
